### PR TITLE
the bugs with func year,month,date,weekday,dayofyear have solved (jus…

### PR DESCRIPTION
**Pull request by JackTan25**
…t think about the result dismatch errors, that's said the expect not equals to result),see #issue3179

**What type of PR is this?**

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #3179 #3269 #3326 #3332

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/sql/plan2/function/builtin/unary/year.go [hunk 1]
```diff
@@ -25,78 +25,103 @@ import (
...
+		// resultValues2 := make([]int64, len(resultValues))
+		// for i, x := range resultValues {
+		// 	resultValues2[i] = int64(x)
+		// }
+		vector.SetCol(resultVector, resultValues)
 		return resultVector, nil
 	}
 }
 
 func DateStringToYear(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
```
Comment (broccoliSpicy) [inline pkg/sql/plan2/function/builtin/unary/year.go:95]: In general, this should be done by implicit casting and we don't need this function

Comment (broccoliSpicy) [reply pkg/sql/plan2/function/builtin/unary/year.go:95]: by casting varchar to date, but I guess it is okay to write this function explicitly

File: pkg/vectorize/year/year.go [hunk 2]
```diff
@@ -60,3 +67,32 @@ func dateStringToYear(xs *types.Bytes, ns *nulls.Nulls, rs []uint16) []uint16 {
...
+	return rs
+}
+
+func dateStringToYearPlan2(xs *types.Bytes, ns *nulls.Nulls, rs []int64) []int64 {
+	for i := range xs.Lengths {
+		str := string(xs.Get(int64(i)))
+		d, e := types.ParseDateCast(str)
+		if e != nil {
+			// set null
+			nulls.Add(ns, uint64(i))
```
Comment (broccoliSpicy) [inline pkg/vectorize/year/year.go:91]: not sure we want to return null when input string is invalid, maybe abort the execution and throw a error?

**Review by broccoliSpicy — APPROVED** [review]

File: pkg/vectorize/year/year.go (hunk 2 above)
Comment (JackTan25) [reply pkg/vectorize/year/year.go:91]: In the test cases, it should be null

Comment (broccoliSpicy) [reply pkg/vectorize/year/year.go:91]: this may need further discussion, I believe the canonical way is to throw error

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**